gateway/ln-gateway/src/ng/register.rs

```diff
@@ -0,0 +1,149 @@
+                register_gateway.transitions(global_context.clone(), self.common.clone())
+            }
+            GatewayRegisterStates::WaitForTTL(wait_for_ttl) => {
+                wait_for_ttl.transitions(self.common.clone())
+            }
+        }
+    }
+
+    fn operation_id(&self) -> fedimint_client::sm::OperationId {
+        self.common.operation_id
+    }
+}
+
+#[derive(Error, Debug, Serialize, Deserialize, Encodable, Decodable, Clone, Eq, PartialEq)]
+pub enum GatewayRegistrationError {
+    #[error("Error registering the Lightning Gateway. Gateway API: {api:?}")]
+    RegistrationError { api: Url },
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct RegisterGateway;
+
+impl RegisterGateway {
+    fn transitions(
+        &self,
+        global_context: DynGlobalClientContext,
+        common: GatewayRegisterCommon,
+    ) -> Vec<StateTransition<GatewayRegisterStateMachine>> {
+        vec![StateTransition::new(
+            Self::await_register_with_federation(global_context, common.clone()),
+            move |_dbtx, res, _| {
+                Box::pin(Self::transition_register_federation(res, common.clone()))
+            },
+        )]
+    }
+
+    async fn await_register_with_federation(
+        global_context: DynGlobalClientContext,
+        common: GatewayRegisterCommon,
+    ) -> Result<(), GatewayRegistrationError> {
+        global_context
+            .module_api()
+            .register_gateway(&common.registration_info)
+            .await
+            .map_err(|_| GatewayRegistrationError::RegistrationError {
+                api: common.registration_info.api,
+            })
+    }
+
+    async fn transition_register_federation(
+        result: Result<(), GatewayRegistrationError>,
+        common: GatewayRegisterCommon,
+    ) -> GatewayRegisterStateMachine {
+        if result.is_err() {
+            error!("{result:?}");
+
+            // Briefly wait and then try again
+            sleep(Duration::from_secs(15)).await;
+            return GatewayRegisterStateMachine {
+                common,
+                state: GatewayRegisterStates::Register(RegisterGateway {}),
+            };
+        }
+
+        GatewayRegisterStateMachine {
+            common,
+            state: GatewayRegisterStates::WaitForTTL(WaitForTimeToLive {}),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct WaitForTimeToLive;
+
+impl WaitForTimeToLive {
+    fn transitions(
+        &self,
+        common: GatewayRegisterCommon,
+    ) -> Vec<StateTransition<GatewayRegisterStateMachine>> {
+        let ttl = common.time_to_live;
+        vec![StateTransition::new(sleep(ttl), move |_, _, _| {
+            Box::pin(Self::transition_wait_for_ttl(common.clone()))
+        })]
+    }
+
+    async fn transition_wait_for_ttl(common: GatewayRegisterCommon) -> GatewayRegisterStateMachine {
+        GatewayRegisterStateMachine {
```

Comment:
Instead of immediately transitioning to re-register, we should probably check the database to see if we should re-register or not

Reply:
Now that we have a metadata database for gatewayd, I think we can just check for the presence of the `FederationId` in the db. Un-registering the federation would just delete that key, then when this workflow wakes up it will see that its not there anymore and transition to `Done` so that it terminates.

Reply:
We now have three cases:

 - `LightningGateway` in the database has been updated with new value -> Transition to `Done`. This means that a new state machine updated the `LightningGateway` and we should just let that one take care of re-registration.
 - `LightningGateway` in the database no longer exists. This means the gateway has been unregistered. Transition to `Done`
 - `LightningGateway` in the database is the same as the previous registration. Transition to `Registering` so that it can be re-registered.